File: frontend/src/api/types/task.ts
export type TaskSourceType = 'requirement' | 'internal' | 'external' | 'adhoc'

export type TaskStatus = 'todo' | 'in_progress' | 'done' | 'cancelled'

export interface TaskAssignee {
  id: number
  username: string
  display_name: string
}

export interface Task {
  id: number
  project_id: number
  requirement_id: number | null
  parent_id: number | null
  title: string
  description: string | null
  source_type: TaskSourceType
  source_description: string | null
  status: TaskStatus
  assignee_id: number | null
  assignee?: TaskAssignee | null
  planned_hours: number | null
  actual_hours: number | null
  cancelled_at: string | null
  cancel_reason: string | null
  created_at: string
  updated_at: string
}

export interface TaskTreeNode extends Task {
  children?: TaskTreeNode[]
}

/** 拆解弹窗 / decompose API 递归输入（仅 title 等编辑字段，无 id） */
export interface DecomposeTaskInput {
  title: string
  description?: string | null
  planned_hours?: number | null
  subtasks?: DecomposeTaskInput[]
}

export interface DecomposePayload {
  tasks: DecomposeTaskInput[]
}

export interface TaskListParams {
  tree?: boolean
  requirement_id?: number
  source_type?: TaskSourceType
  skip?: number
  limit?: number
}

export interface TaskListResponse {
  items: Task[]
  total: number
  skip: number
  limit: number | null
}

export interface TaskTreeListResponse {
  items: TaskTreeNode[]
  total: number
  skip: number
  limit: number | null
}

export interface TaskCreatePayload {
  title: string
  description?: string | null
  requirement_id?: number | null
  parent_id?: number | null
  source_type: TaskSourceType
  source_description?: string | null
  assignee_id?: number | null
  planned_hours?: number | null
  status?: TaskStatus
}

export interface TaskUpdatePayload {
  title?: string
  description?: string | null
  parent_id?: number | null
  assignee_id?: number | null
  planned_hours?: number | null
  status?: TaskStatus
}

export interface TaskCancelPayload {
  cancel_reason: string
}

export const TASK_SOURCE_TYPE_LABELS: Record<TaskSourceType, string> = {
  requirement: '需求拆解',
  internal: '项目内',
  external: '项目外',
  adhoc: '临时任务',
}

export const TASK_STATUS_LABELS: Record<TaskStatus, string> = {
  todo: '待办',
  in_progress: '进行中',
  done: '已完成',
  cancelled: '已取消',
}

File: backend/app/schemas/task.py
from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import List, Optional

from pydantic import BaseModel, ConfigDict, Field, field_serializer

from app.models.task import TaskSourceType, TaskStatus


class TaskCreate(BaseModel):
    title: str = Field(min_length=1, max_length=200)
    description: Optional[str] = None
    requirement_id: Optional[int] = None
    parent_id: Optional[int] = None
    source_type: TaskSourceType
    source_description: Optional[str] = None
    assignee_id: Optional[int] = None
    planned_hours: Optional[float] = Field(default=None, ge=0)
    status: TaskStatus = TaskStatus.TODO


class TaskUpdate(BaseModel):
    title: Optional[str] = Field(default=None, min_length=1, max_length=200)
    description: Optional[str] = None
    parent_id: Optional[int] = None
    assignee_id: Optional[int] = None
    planned_hours: Optional[float] = Field(default=None, ge=0)
    status: Optional[TaskStatus] = None


class DecomposeTaskInput(BaseModel):
    title: str = Field(min_length=1, max_length=200)
    description: Optional[str] = None
    planned_hours: Optional[float] = Field(default=None, ge=0)
    subtasks: List["DecomposeTaskInput"] = []


class DecomposePayload(BaseModel):
    tasks: List[DecomposeTaskInput] = Field(min_length=1)


class TaskCancel(BaseModel):
    cancel_reason: str = Field(min_length=1)


class TaskAssigneeBrief(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    username: str
    display_name: str


class TaskResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    project_id: int
    requirement_id: Optional[int]
    parent_id: Optional[int]
    title: str
    description: Optional[str]
    source_type: TaskSourceType
    source_description: Optional[str]
    status: TaskStatus
    assignee_id: Optional[int]
    assignee: Optional[TaskAssigneeBrief] = None
    planned_hours: Optional[float] = None
    actual_hours: Optional[float] = None
    cancelled_at: Optional[datetime]
    cancel_reason: Optional[str]
    created_at: datetime
    updated_at: datetime

    @field_serializer("planned_hours", "actual_hours")
    def serialize_hours(self, value: Optional[Decimal]) -> Optional[float]:
        if value is None:
            return None
        return float(value)


class TaskTreeNode(TaskResponse):
    children: List[TaskTreeNode] = []


class TaskListResponse(BaseModel):
    items: List[TaskResponse]
    total: int
    skip: int
    limit: Optional[int]


class TaskTreeListResponse(BaseModel):
    items: List[TaskTreeNode]
    total: int
    skip: int
    limit: Optional[int]


DecomposeTaskInput.model_rebuild()
TaskTreeNode.model_rebuild()
